### PR TITLE
Added the device_tokens endpoint

### PR DESCRIPTION
**Pull request by mllocs**
In order to send push notifications we need a place to store the device tokens for each users. In this PR:

- I'm adding a new table (`device_tokens`) with a minimal API (just the create action).
- A couple of services and classes to send push notifications.

The future TimeOverflow mobile app will be implemented in this repo: https://github.com/coopdevs/timeoverflow-mobile-app 👶📱

### CODE REVIEW

File: app/controllers/device_tokens_controller.rb [hunk 1]
```diff
@@ -0,0 +1,19 @@
+class DeviceTokensController < PostsController
+  before_filter :authenticate_user!
+
+  def create
+    @device_token = DeviceToken.all.build device_token_params.merge! user_id: current_user.id
```
[the remark this answers is not in the document]
Comment (sauloperez) [reply app/controllers/device_tokens_controller.rb]: It's enough with a `DeviceToken.new(...)`, isn't it?

File: app/services/push_notifications/expo_adaptor_service.rb [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+      end
+    end
+
+    private
+
+    def post_data(token)
+      {
+        "to" => token,
+        "title" => @notification.title,
+        "body" => @notification.body
```
Comment (sauloperez) [inline app/services/push_notifications/expo_adaptor_service.rb]: Please, use attr_readers for the ivars

File: app/services/push_notifications/post_broadcaster_service.rb [hunk 3]
```diff
@@ -0,0 +1,15 @@
+module PushNotifications
+  class PostBroadcasterService
+    def initialize(post)
+      @post = post
+    end
+
+    def broadcast
+      notification = PostNotification.new
+
+      @post.organization.users.each do |user|
```
Comment (sauloperez) [inline app/services/push_notifications/post_broadcaster_service.rb]: I would decouple this class from the details of the data model by passing the list of users in the constructor instead. This way if we ever change the data model we don't need to touch the push notifications implementation.

File: db/migrate/20180221161343_create_device_tokens.rb [hunk 4]
```diff
@@ -0,0 +1,12 @@
+class CreateDeviceTokens < ActiveRecord::Migration
+  def change
+    create_table :device_tokens do |t|
+      t.integer :user_id, :null => false
+      t.string :token, :null => false
+
+      t.timestamps
+    end
+    add_index :device_tokens, [:user_id, :token]
+    add_index :device_tokens, [:user_id]
```
Comment (sauloperez) [inline db/migrate/20180221161343_create_device_tokens.rb]: This one is useless due to the left-most thingie. Postgresql can use that one above when only needing user_id

File: app/services/push_notifications/post_broadcaster_service.rb [hunk 5]
```diff
@@ -0,0 +1,19 @@
+module PushNotifications
+  class UsersBroadcasterService
```
Comment (sauloperez) [inline app/services/push_notifications/post_broadcaster_service.rb]: you're still working on it, right? because I don't see this class used.

Comment (mllocs) [reply app/services/push_notifications/post_broadcaster_service.rb]: this class will be used once we implement the logic behind sending push notifications to users :)

File: db/migrate/20180221161343_create_device_tokens.rb [hunk 6]
```diff
@@ -0,0 +1,11 @@
+class CreateDeviceTokens < ActiveRecord::Migration
+  def change
+    create_table :device_tokens do |t|
+      t.integer :user_id, :null => false
+      t.string :token, :null => false
+
+      t.timestamps
+    end
+    add_index :device_tokens, [:user_id, :token]
```
Comment (mllocs) [inline db/migrate/20180221161343_create_device_tokens.rb]: this should be uniq

File: db/migrate/20180221161343_create_device_tokens.rb [hunk 7]
```diff
@@ -0,0 +1,11 @@
+class CreateDeviceTokens < ActiveRecord::Migration
+  def change
+    create_table :device_tokens do |t|
+      t.integer :user_id, :null => false
+      t.string :token, :null => false
+
+      t.timestamps
+    end
+    add_index :device_tokens, [:user_id, :token], unique: true
```
Comment (sauloperez) [inline db/migrate/20180221161343_create_device_tokens.rb]: :clap:

File: app/controllers/device_tokens_controller.rb [hunk 8]
```diff
@@ -0,0 +1,19 @@
+class DeviceTokensController < PostsController
```
Comment (enricostano) [inline app/controllers/device_tokens_controller.rb]: Why is this controller inheriting from `PostsController`?

Comment (sauloperez) [reply app/controllers/device_tokens_controller.rb]: good catch

Comment (mllocs) [reply app/controllers/device_tokens_controller.rb]: waaaaaaat 👓

**Comment by sauloperez** [comment]
If I understand this correctly we're still missing 

> A couple of services and classes to send push notifications.

right? In that case I would remove the route from this PR, merge and iterate on it an upcoming PR. It feels like all we can do as a first step for push notifications is done here.

**Comment by mllocs** [comment]
Yup, I would also merge this, even if we there's nothing that sends notifications. I wouldn't remove the route though, since we could start registering mobile devices.

**Pull request merged**